aiohttp/client_proto.py

```diff
@@ -234,14 +271,22 @@ def data_received(self, data: bytes) -> None:
                 # parse http messages
                 try:
                     messages, upgraded, tail = self._parser.feed_data(data)
-                except BaseException as exc:
+                except BaseException as underlying_exc:
                     if self.transport is not None:
                         # connection.release() could be called BEFORE
                         # data_received(), the transport is already
                         # closed in this case
                         self.transport.close()
                     # should_close is True after the call
-                    self.set_exception(exc)
+                    reraised_exc = ClientPayloadError(
+                        f"Unable to parse response payload: {underlying_exc !r}"
+                    )
+                    if isinstance(
+                        underlying_exc,
+                        (ClientError, HttpProcessingError),
+                    ):
+                        reraised_exc = underlying_exc
```

Comment:
@Dreamsorcerer I did this to fix the tests. But what do you think about changing this to a different error unconditionally?

I'm a bit concerned about the changes in the PR influencing the end-users...

Reply:
We've already been told by users to stop throwing exceptions that are not inhereting from ClientError. They expect to be able to catch all exceptions with that, so this is probably the desired behaviour.

Reply:
Yes, but this hack makes the underlying error leak/bubble up directly to the end users, which means the traceback won't have this layer, right?

Reply:
I'm not sure, but I suspect it won't make a noticeable difference to users. Maybe just skip the 3.9 backport to minimise impact?

Reply:
@Dreamsorcerer do you mean keep this the logic in the PR as it is now, backport it to 3.9 and 3.10, and only change the logic in a separate PR targeting 3.10?
I definitely want to use this patch to discover what's happening with my github bots with aiohttp 3.9 and not wait till 3.10.

I'll see if there's a meaningful way to make a patch that migrates to throwing something inheriting from `ClientError` while declaring it a bugfix.

@bdraco do you have any thoughts on this?

Reply:
I don't mind either way. If you think it's likely to cause some minor breakage, it'd make sense to delay it to 3.10. But, I have no feeling for this, so I'm fine with it going into 3.9 if you are.

Reply:
I think it seems reasonable as a bugfix to make it always inherit from `ClientError`.  I expect consumers already expect that given https://github.com/aio-libs/aiohttp/pull/8089/files#r1471738716

Reply:
So it occurred to me that the protocol is an internal layer, probably. And maybe, it should keep the internal exceptions as is with them being re-raised on the outer layer. Though, I'm yet to look deeper into it — this is the last thing keeping me from merging the patch.